### PR TITLE
catch common resource schema issues in cfn validate

### PR DESCRIPTION
**Pull request by PatMyron**
continuing https://github.com/aws-cloudformation/cloudformation-cli/pull/663, https://github.com/aws-cloudformation/cloudformation-cli/pull/668
fixes https://github.com/aws-cloudformation/cloudformation-cli/issues/459, https://github.com/aws-cloudformation/cloudformation-cli/issues/414

---

**[how to run new validations on all existing resource provider schemas](https://github.com/aws-cloudformation/cloudformation-cli/pull/604#issuecomment-756521889)**

---

* [incorrect min/max constraints](https://github.com/aws-cloudformation/cloudformation-cli/issues/414)
* [inconsistent property casing](https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/pull/77)
```bash
curl -s -N --compressed https://d1uauaxba7bl26.cloudfront.net/latest/gzip/CloudFormationResourceSpecification.json | grep -E '"[a-g][i-z]'
        "error": {
        "code": {
```
* [hardcoded aws partition patterns](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-syntax)
```bash
CloudformationSchemas $ grep 'pattern.*arn:aws:' *
aws-appflow-flow.json:      "pattern" : "arn:aws:appflow:.*:[0-9]+:.*",
aws-appflow-flow.json:      "pattern" : "arn:aws:kms:.*:[0-9]+:.*",
aws-globalaccelerator-endpointgroup.json:      "pattern" : "^arn:aws:(\\w+)::(\\d{12}):(accelerator)/([0-9a-f-]+)/listener/([0-9a-f-]+)$"
aws-ivs-channel.json:      "pattern" : "^arn:aws:ivs:[a-z0-9-]+:[0-9]+:channel/[a-zA-Z0-9-]+$",
aws-ivs-playbackkeypair.json:      "pattern" : "^arn:aws:ivs:[a-z0-9-]+:[0-9]+:playback-key/[a-zA-Z0-9-]+$",
aws-ivs-streamkey.json:      "pattern" : "^arn:aws:ivs:[a-z0-9-]+:[0-9]+:stream-key/[a-zA-Z0-9-]+$",
aws-ivs-streamkey.json:      "pattern" : "^arn:aws:ivs:[a-z0-9-]+:[0-9]+:channel/[a-zA-Z0-9-]+$"
aws-mwaa-environment.json:      "pattern" : "^(((arn:aws(-[a-z]+)?:kms:[a-z]{2}-[a-z]+-\\d:\\d+:)?key\\/)?[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}|(arn:aws:kms:[a-z]{2}-[a-z]+-\\d:\\d+:)?alias/.+)$"
aws-opsworkscm-server.json:      "pattern" : "arn:aws:iam::[0-9]{12}:role/.*",
aws-opsworkscm-server.json:      "pattern" : "arn:aws:iam::[0-9]{12}:instance-profile/.*",
aws-sso-assignment.json:      "pattern" : "arn:aws:sso:::instance/(sso)?ins-[a-zA-Z0-9-.]{16}",
aws-sso-assignment.json:      "pattern" : "arn:aws:sso:::permissionSet/(sso)?ins-[a-zA-Z0-9-.]{16}/ps-[a-zA-Z0-9-./]{16}",
aws-sso-instanceaccesscontrolattributeconfiguration.json:      "pattern" : "arn:aws:sso:::instance/(sso)?ins-[a-zA-Z0-9-.]{16}",
aws-sso-permissionset.json:      "pattern" : "arn:aws:sso:::permissionSet/(sso)?ins-[a-zA-Z0-9-.]{16}/ps-[a-zA-Z0-9-./]{16}",
aws-sso-permissionset.json:      "pattern" : "arn:aws:sso:::instance/(sso)?ins-[a-zA-Z0-9-.]{16}",
```
* [invalid regular expression patterns](https://github.com/aws-cloudformation/cloudformation-cli/issues/459)
* [non-ASCII resource schema characters](https://github.com/aws-cloudformation/cfn-python-lint/pull/1867#discussion_r562232974)
* [hardcoding constantly evolving enums like instance types, lambda runtimes, partitions, regions, availability zones, etc. that get outdated quickly](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-route53/pull/24)

### CODE REVIEW

File: src/rpdk/core/data_loaders.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import json
 import logging
 import os
+import re
```
Comment (kddejong) [inline src/rpdk/core/data_loaders.py:4]: I like this type of testing.  Get rid of some of the java patterns would be helpful. My question is the python re the one we want to standardize on?  For instance `^[\d\w-_.+]*$` is technically valid but in python it needs to be `^[\d\w\-_.+]*$`.

Comment (PatMyron) [reply src/rpdk/core/data_loaders.py:4]: Ideally, resource schema patterns are valid and equivalent in both Python and Java

[JSON schema itself recommends sticking to a minimal subset of regular expression syntax](http://json-schema.org/understanding-json-schema/reference/regular_expressions.html). Let's encourage the same

Just emitting warnings for patterns not valid in Python seems like a good start in that direction

Comment (kddejong) [reply src/rpdk/core/data_loaders.py:4]: > JSON schema itself recommends sticking to a minimal subset of regular expression syntax. Let's encourage the same

Agreed completely here.

**Review by jotompki — COMMENTED** [review]
I would advise trying to use the schema flattener to check any nested object's properties. The flattener will place all nested schemas in the schema_map for use. iterating over these as well as the top level definitions and properties should allow you to definitively check all objects

File: src/rpdk/core/data_loaders.py [hunk 2]
```diff
@@ -141,6 +143,111 @@ def load_resource_spec(resource_spec_file):  # pylint: disable=R0912 # noqa: C90
...
+        "maxLength",
+        "minProperties",
+        "maxProperties",
+        "minItems",
+        "maxItems",
+        "exclusiveMinimum",
+        "exclusiveMaximum",
+    }
+
+    def check_property(property_name, property_details):
```
[the remark this answers is not in the document]
Comment (PatMyron) [reply src/rpdk/core/data_loaders.py:159]: Can't imagine it being useful anywhere else

**Review by omkhegde — APPROVED** [review]
LGTM, I suggest adding a test case with a large schema file with a lot of nested properties.

**Review by jotompki — APPROVED** [review]

**Comment by PatMyron** [comment]
> LGTM, I suggest adding a test case with a large schema file with a lot of nested properties.

ran `cfn validate` on all the [existing resource schemas](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-type-schemas.html), which should have quite a few nested properties

File: src/rpdk/core/data_loaders.py (hunk 1 above)
Comment (PatMyron) [reply src/rpdk/core/data_loaders.py:4]: Golang has some interesting regex limitations even within that minimal subset:
https://github.com/hashicorp/terraform-provider-awscc/issues/88
https://github.com/golang/go/issues/7252

**Pull request merged**